src/python/pants/core/util_rules/source_files.py

```diff
@@ -31,17 +51,65 @@ class SourceFilesRequest:
     sources_fields: Tuple[SourcesField, ...]
     for_sources_types: Tuple[Type[SourcesField], ...]
     enable_codegen: bool
+    ignore_unparented_assets: bool
```

Comment:
I think that the fact that this is a double negative has me legitimately confused about the behavior... the inverted name would be something like `require_parented_assets` with a default of `False`?

As you mention in the description, there is code elsewhere which fails for this case. Should this flag instead be `unparented_assets_behavior={Fail,Warn,Ignore}`? It would help put it in context to actually use it somewhere.

Reply:
`require_parented_assets` wouldn't quite be it because we aren't _requiring_ anything, just tossing some things out the window.

Additionally, nothing fails today. It either ignores, ignore+warns, or doesn't think about it.

I'll see if I can introduce the use-cases in this PR.